### PR TITLE
Alternative Syntax for Control Expressions

### PR DESCRIPTION
**Pull request by odersky**
This PR implements a new "quiet" syntax for control expressions that does not rely in
enclosing the condition in parentheses, and also allows to drop parentheses or braces
around the generators of a `for`-expression. Examples:
```scala
if x < 0 then -x else x

while x >= 0 do x = f(x)

for x <- xs if x > 0
yield x * x

for
  x <- xs
  y <- ys
do
  println(x + y)
```

The rules in detail are:

 - The condition of an `if`-expression can be written without enclosing parentheses if it is followed by a `then`.
 - The condition of a `while`-loop can be written without enclosing parentheses if it is followed by a `do`.
 - The enumerators of a `for`-expression can be written without enclosing parentheses or braces if they are followed by a `yield` or `do`.
 - A `do` in a `for`-expression expresses a `for`-loop.
 - Newline characters are not statement separators in a condition of an `if` or a `while`.
   So the meaning of newlines is the same no matter whether parentheses are present
   or absent.
 - Newline characters are statement separators in the enumerators of a `for`-expression. This turns out to be not a problem because #7031 basically removed the need to put expressions in parentheses to get nice operator formatting.

### Rewrites

The Dotty compiler can rewrite source code from old syntax and new syntax and back.
When invoked with options `-rewrite -new-syntax` it will rewrite from old to new syntax, dropping parentheses and braces in conditions and enumerators. When invoked with with options `-rewrite -old-syntax` it will rewrite in the reverse direction, inserting parentheses and braces as needed.

### CODE REVIEW

**Comment by jducoeur** [comment]
Pure opinion: I really dislike this sort of syntax -- I find it significantly harder to scan code and understand the control flow in keyword-centric languages like this.  It's the sort of thing that would actively nudge me away from Scala.

I get that it's just being presented as an option, and that's fine in principle, but I gotta say: it seems contrary to the general bent of the Scala 3 project, of being a bit more opinionated about style.  I would probably forbid it in any codebase I'm leading...

**Comment by sjrd** [comment]
I am very concerned about this. I grew up with Pascal/Delphi, so I'm not against the new syntax *per se*. I am concerned about the *change*.

The issue is that it will inevitably lead to a fracture of the ecosystem. A lot of code will use the old syntax, including all the existing material in books and on the Web. And supposedly a lot of new code will use the new syntax. We'll have zillions of people wondering how to read other people's code, or if the two things are the same, or which one they should use, etc.

We've seen this problem before for other things, for example `implicit`. However in this case we're talking about *an entirely different scale*: every single Scala file will be affected! From a beginner's code to an expert's. And for what? A purely syntactic change. The ratio benefit/migration pain will be extremely low for this change. Much lower than other changes we've been seeing in Scala 3.

The low ratio is not helped by the fact that I see little in the way of motivation for this change. With a historical record of everything that's been discussed the past few years, I can see this being motivated in the context of indentation-based syntax, which is a radically different syntax aspect. With such a change, there are objective reasons to switch to `if then` and `while do`, notably the fact that, otherwise, there is no way to write multi-statement conditions without introducing braces in a world that is otherwise brace-free.

In the context of brace-based syntax, I could see a much more restrictive version of this proposal: allow braces instead of parens for the condition of `while` and `if`. That would bring them in line with `for` (for which we have the choice of `{}` or `()`) and would avoid the argued-as-ugly `({ ... })`, allowing to write:
```scala
while {
  val c = str.charAt(i)
  c >= '0' && c <= '9'
} {
  i += 1
}
```

---

Aside:

Speaking of indentation-based syntax, using it would basically require treating newlines in conditions as statement separators, so that I could for example write the following:
```scala
while
  val c = str.charAt(i)
  c >= '0' && c <= '9'
do
  ...
```
If we don't treat them as newlines now, then treating them as newlines with indentation syntax would be an even more problematic change. For that reason, at least newlines should be treated as statement separators now (and hence systematically fail to parse, since we are not in a block under non-indentation-based syntax), if there is still any ounce of desire to have indentation-based syntax in the future.

**Comment by odersky** [comment]
@jducoeur @sjrd This is not intended to stay an alternative for long. My current thinking is that we should standardize on the new syntax, and rewrite all old syntax automatically (this PR proves that this is feasible). Using indentation or not is a separate aspect. I am ready argue that the new syntax is vastly superior even if we stick with braces.

**Comment by hrhino** [comment]
What's changed since [last time](https://docs.scala-lang.org/sips/minutes/2016-08-16-sip-10th-august-minutes.html)?

> allow braces instead of parens for the condition of `while` and `if`

This is a nice tweak, and I'd hope it could make it in. Notably, it doesn't require rewriting existing code, nor would such a rewrite rule be widely applicable.

I do like the `while { ... } do { ... }` syntax, probably because adjacent close/open braces look grating to me.

**Comment by propensive** [comment]
I'm cautiously, hesitantly willing to accept this new style once I remind myself that *it's only syntax*. I've no doubt we'll all get used to it.

However, one minor benefit of the current brace-based syntax is that text editors and IDEs can very easily match an opening to a closing brace, and indicate this to the user.

**Comment by rubenfiszel** [comment]
I for one really like the added symmetry and welcome the change

**Comment by molikto** [comment]
I don't think rewrite tools solves all problems, there are people generating Scala code automately, or projects uses modified Scala syntax (https://github.com/lihaoyi/Scalatex), also all the code snippets written in various blog posts, forums on the web. It will just be a lot of pain.

I wonder if instead we can do a structural editor with configurable syntax presentation (I am interested in experiment with this).

**Comment by odersky** [comment]
Rebased on top of #7031

**Comment by odersky** [comment]
Squashed to a single commit on top of #7024

**Comment by jdegoes** [comment]
Very beautiful improvement to the syntax of Scala.

This streamlined syntax emphasizes the expression-oriented nature of Scala (reminiscent of the change in method syntax,  `def foo: Unit { }` vs `def foo: Unit = ???`). Expression-orientation, and more generally, functional programming, are among the strengths of Scala; so I personally welcome the change.

I do think rewriting and deprecation of the old syntax are both important for any change of this magnitude: rewriting to ease migration, deprecation to enforce a single style in large code bases.

Now I know some will object to this change, on grounds it makes Scala a worse _Better Java_. There is a kernel of truth here; those who love curly braces will find more streamlined syntax yet another reason to stick with Java or Kotlin. But I would suggest that this change makes Scala a better _Scala_, and that the _Better Java_ market is not a growth market for Scala.

Python has also shown, with its tremendous success, that the lack of curly braces or the presence of indentation-sensitive syntax are no impediments to adoption (indeed, most languages would be lucky to have a fraction of Python's success).

In summary, this change is a great example of Scala "doubling-down" on Scala—a change courageously opinionated enough to (yes) give some another reason to stay with Java, but also bold enough to galvanize those who love the conciseness and expressive power of Scala, and who would love to see a change this impactful to end-user ergonomics and aesthetics.

**Comment by jducoeur** [comment]
Just to be clear: I intensely dislike Java.  Indeed, that was what drove me to learn Scala back in 2007.  I'm not longing for better-Java-style Scala here.

My preference for the braces has nothing to do with that -- it has to do with me finding indentation-centric code just plain harder to scan when functions gets beyond about 10 lines (which in practice happens often, however much we might wish otherwise, especially when you have developers with wildly different experience levels), and prone to errors during refactoring if you're not careful to realign things to suit the environment.  I strongly suspect that I would find it unusable outside of an IDE.

So I sincerely believe that the braceless syntax will make my codebases harder to *maintain* in the long run, and very little is higher priority to me than that.  Between that and the fact that the dramatic visual difference is going to make the retraining effort much more obvious, I am (quite legitimately, I think) worried that I will have a much harder time getting my company (a relatively large one as Scala-centric firms go) to upgrade.  IMO, the supposed benefits of this change are nowhere near the costs...

**Comment by diesalbla** [comment]
>  The Dotty compiler can rewrite source code from old syntax and new syntax and back.

As a believer in ["_Make each program do one thing well_"](http://www.catb.org/~esr/writings/taoup/html/ch01s06.html), I would avoid adding any code rewrite features to a compiler, and leave that to a separate tool like `scalafmt`.

**Comment by ryanberckmans** [comment]
@odersky might we support golang-style control flow syntax, which is omission of conditional parentheses but still braces around the block? 

(I have limited interest in significant indentation syntax and will most likely keep using curly braces. But I find your exploration of "radical" changes to be inspiring and will give it a try, especially given the excellent tool support to convert sources :-)

Golang style might look like

```scala
def hello(x: Int): Unit = if x > 5 { 
  println(x)
  println(x*2)
}
```

**Comment by som-snytt** [comment]
I love this change. It's like those mornings I remember to clean my eyeglasses. The rest of the day is better and clearer.

However, for a very long time, I've been accustomed to vi getting confused by stray braces in comments and then failing to match braces correctly, so I'll have to get used to not dealing with that issue.

**Comment by jxtps** [comment]
IFF we go full indentation-crazy, how about allowing the user to skip the new keywords and instead literally double down on indentation: use an extra indentation to denote what's inside the missing parenthesis: 

```scala
if x < 0 && 
    y > 1  // <-- extra indent means it's "inside the ()"
  -x 
else 
  x

while x >= 0 && 
    y > 1 // <-- extra indent means it's "inside the ()"
  x = f(x)

for x <- xs 
    if x > 0 // <-- extra indent means it's "inside the ()"
yield 
  x * x

for x <- xs
    y <- ys // <-- extra indent means it's "inside the ()"
  println(x + y)
```

Maybe it's a little better if the "inside the ()" part lines up with the keyword+space: 

```scala
if x < 0 && 
   y > 1  // <-- extra indent means it's "inside the ()"
  -x 
else 
  x

while x >= 0 && 
      y > 1 // <-- extra indent means it's "inside the ()"
  x = f(x)

for x <- xs 
    if x > 0 // <-- extra indent means it's "inside the ()"
yield 
  x * x

for x <- xs
    y <- ys // <-- extra indent means it's "inside the ()"
  println(x + y)
```
Maybe too subtle for the `if` since it's only one space off from the regular code? A subtle-ish fix for that would be to introduce an extra space after the `if` when there's multi-line logic (yes, that has a host of issues): 

```scala
if  x < 0 && 
    y > 1  // <-- extra indent means it's "inside the ()"
  -x 
else 
  x
```


Not sure I like it, but wanted to put it out there as it's not going to get much lighter than this, and I hadn't seen this in any of the significant indentation discussions.

**Comment by texasbruce** [comment]
This got me thinking - I feel like we are heading to the wrong direction.

To get Scala more widely used, is changing syntax the correct way? Scala needs a killer use like other popular languages (Rails for Ruby, ReactNative for JS, Android Dev for Kotlin, etc.), and, to me, syntax is not the road blocker for getting people to use it.

**Comment by ryanberckmans** [comment]
@texasbruce 

> To get Scala more widely used, is changing syntax the correct way?

Yes, this is an important question. There may be a better place to discuss this, but let's start here.

In my opinion the main blocker to Scala's adoption (besides a reputation for brutal compile times) is the pervasive _conventions_ that define idiomatic Scala usage.

I wrote about this https://contributors.scala-lang.org/t/principles-for-implicits-in-scala-3/3072/51

> if typeclasses are such a powerful, obvious, necessary part of Scala, why is there no Monoid in the standard library? 

I think "Step 1a" to improving mass adoption of Scala is to merge cats into the standard library.

**Comment by nafg** [comment]
To me the question is, since cannot predict but only speculate on the future, what is the potential gain, and what is the potential risk.

For example, there could be a scenario where due to changed syntax scala becomes X% more popular, but there is another conceivable scenario where it results in a total of Y% _less_ popular. What are values for X and Y that we should think about?

**Comment by jackmaney** [comment]
> To get Scala more widely used, is changing syntax the correct way?

Looking from the outside in, the Scala community is--to put it as politely as I can--a dumpster fire. Fixing that will have a much better effect than any syntax changes.

**Comment by ryanberckmans** [comment]
@jackmaney I'd just like to offer a counter perspective in that, being relatively new to Scala, I've been pleasantly delighted by the language, standard library, 3rd party libraries, helpful Q&A on reddit, community members with whom I've interacted, etc.

I am particularly impressed by the thoughtfulness and execution so far on the transition to Scala 3. The talks from Scala Days 2019 were great. The communication around Dotty is great. The Dotty website is great.

**Comment by leobenkel** [comment]
Having the indentation part of the syntax is a terrible idea. 

Code style and code meaning should be two distinct things. 

Also, I like my curly to eye ball where is the start and end of blocks. 
You can put your cursor on one and instantly know where the other end is. 
You can copy paste code without breaking its functionality. 

Have you even copy paste Python code from one place to another ? It is a nightmare.

**Comment by akilegaspi** [comment]
I don't think people are reading enough.

It is an optional syntax hence the "Alternative" in the title, you may or may not want to use it depending on your preference and a way around it is to enforce a format in your coding shop that doesn't use this syntax if you don't want it to be a plague in your codebase.

**Comment by philliptaylorpro** [comment]
As a former Perl developer I think this serves no purpose other than to create more syntax to learn before you can be proficient in Scala and both slows you down and knocks your confidence as you read other people's code.
